File: psoscheduler/src/main/java/syslab/cloudcomputing/pso/Particle.java
package syslab.cloudcomputing.pso;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

import syslab.cloudcomputing.simulation.DataCenter;
import syslab.cloudcomputing.simulation.Task;
import syslab.cloudcomputing.simulation.VirtualMachine;
import syslab.cloudcomputing.simulation.Workload;
import syslab.cloudcomputing.utils.Matrix;
import syslab.cloudcomputing.utils.Utilities;

/*
 * Particle represents a complete solution to the optimization problem, meaning a complete 
 * mapping of tasks to virtual machines. Each particle's goal is to move towards the
 * optimized mapping of tasks to virtual machine by maximizing the objective function
 */
public class Particle {
  final static double c1 = 2.0;
  final static double c2 = 1.49455;

  final static double maxAbsoluteVelocity = 10.0;  // from https://www.sciencedirect.com/science/article/pii/S1319157820305279#e0045

  private static int idCounter = 1;
	private int id;

  // The velocity represents a 2D array where the rows are the tasks and the columns are the positions
  // Source: https://www.sciencedirect.com/science/article/pii/S1319157820305279
  private Matrix position;
  private Matrix velocity;
  private HashMap<Task, VirtualMachine> taskVmMapping = new HashMap<Task, VirtualMachine>();

  private Matrix personalBestPosition;
  private double personalBestObjectiveValue;
  private double personalBestMakespan;
  private double personalBestThroughput;

  private ArrayList<Double> objectiveHistory = new ArrayList<Double>();

  private Matrix globalBestPosition;

  private Workload workload;
  private DataCenter dataCenter;

  public static enum InitializationStrategy {
    RANDOM,
    MCT,
    HIGH_TASK_HIGH_VM
  }

  public Particle(DataCenter dataCenter, Workload workload) {
    this.id = Particle.idCounter++;
    this.workload = workload;
    this.dataCenter = dataCenter;
    this.randomInitialization();
  }

  public Particle(DataCenter dataCenter, Workload workload, InitializationStrategy initializationStrategy) {
    this.id = Particle.idCounter++;
    this.workload = workload;
    this.dataCenter = dataCenter;
    if (initializationStrategy == Particle.InitializationStrategy.MCT) {
      this.mctInitialization();
    } else if (initializationStrategy == Particle.InitializationStrategy.HIGH_TASK_HIGH_VM) {
      this.highTaskHighVmInitialization();
    } else {
      this.randomInitialization();
    }
  }

  // Returns true is this iteration was a success (ie. it acheived a new personal best)
  public int runIteration(double w) {

    this.updateVelocity(w);
    this.updatePosition();

    this.updateDataCenter();
    double objective = this.dataCenter.computeObjective();

    this.objectiveHistory.add(this.personalBestObjectiveValue);

    if (objective > this.personalBestObjectiveValue) {
      // System.out.println("BETTER POSITION");
      this.personalBestPosition = this.position.copy();
      this.personalBestObjectiveValue = objective;
      // TODO remove the following two lines when benchmarking
      this.personalBestMakespan = this.dataCenter.computeMakespan();
      this.personalBestThroughput = this.dataCenter.computeThroughput();
      return 1;
    }
    // } else if (objective < this.personalBestObjectiveValue){
    //   System.out.println("WORSE POSITION");
    // } else {
    //   System.out.println("SAME POSITION");
    // }

    return 0;
  }

  private void randomInitialization() {
    this.position = new Matrix(this.workload.getTaskCount(), this.dataCenter.getVirtualMachineCount());

    this.velocity = new Matrix(this.workload.getTaskCount(), this.dataCenter.getVirtualMachineCount());
    this.velocity.randomVelocityInitialization(0, 1);

    this.updatePosition();

    updateDataCenter();

    // Must compute the local best after building the task to vm mapping in the line above for accurate
    // objective function computation
    this.personalBestPosition = this.position;
    this.personalBestObjectiveValue = this.dataCenter.computeObjective();
    // TODO remove
    this.personalBestMakespan = this.dataCenter.computeMakespan();
    this.personalBestThroughput = this.dataCenter.computeThroughput();
  }

  // Minimum completion time algorithm used for initialization as per https://www.sciencedirect.com/science/article/pii/S1319157820305279#e0045
  private void mctInitialization() {
    // Attempt to overcome this issue with getting stuck in local optimums: https://www.mdpi.com/1999-4893/11/2/23
    // Potential remedy: https://www.intechopen.com/journals/1/articles/117
    this.position = new Matrix(this.workload.getTaskCount(), this.dataCenter.getVirtualMachineCount());

    this.dataCenter.resetVirtualMachineReadyTimes();

    ArrayList<Task> tasks = new ArrayList<Task>(this.workload.getTasks());
    Collections.shuffle(tasks);

    for (Task t : tasks) {
      VirtualMachine vm = this.dataCenter.getVmWithMinEET(t);
      if (Utilities.getRandomDouble(0, 1) > 0.3) {
        this.position.setComponent(t.getId() - 1, vm.getId() - 1, 1);
        this.dataCenter.addExecutionTimeToVirtualMachine(t, vm);
      } else {
        int vmId = Utilities.getRandomInteger(0, this.dataCenter.getVirtualMachineCount() - 1);
        VirtualMachine randomVm = this.dataCenter.getVirtualMachineById(vmId);
        this.position.setComponent(t.getId() - 1, vmId, 1);
        this.dataCenter.addExecutionTimeToVirtualMachine(t, randomVm);
      }
    }
    
    this.velocity = new Matrix(this.workload.getTaskCount(), this.dataCenter.getVirtualMachineCount());
    this.velocity.randomVelocityInitialization(0, 1);

    updateDataCenter();
    this.personalBestPosition = this.position;
    this.personalBestObjectiveValue = this.dataCenter.computeObjective();
    // TODO remove
    this.personalBestMakespan = this.dataCenter.computeMakespan();
    this.personalBestThroughput = this.dataCenter.computeThroughput();
  }

  // Initialization strategy used in this paper: https://www.mdpi.com/2079-9292/12/12/2580
  // Top 25% of tasks by MI assigned to Virtual Machines with higher processing capacity
  // The rest follow a random allocation strategy
  private void highTaskHighVmInitialization() {
    this.position = new Matrix(this.workload.getTaskCount(), this.dataCenter.getVirtualMachineCount());
    
    ArrayList<VirtualMachine> sortedVms = this.dataCenter.getSortedVirtualMachines();
    ArrayList<Task> sortedTasks = this.workload.getSortedTasks();

    int vmQ3 = sortedVms.size() * 1 / 2;
    int taskQ3 = sortedTasks.size() * 1 / 2;
    
    for (int i = 0; i < sortedTasks.size(); i++) {
      VirtualMachine vm;
      Task t = sortedTasks.get(i);
      if (i < taskQ3) {
        // Assign to low capacity machine
        vm = sortedVms.get(Utilities.getRandomInteger(0, vmQ3));
      } else {
        // Assign to high capacity machine
        vm = sortedVms.get(Utilities.getRandomInteger(vmQ3, sortedVms.size()));
      }
      this.position.setComponent(t.getId() - 1, vm.getId() - 1, 1);
    }

    this.velocity = this.position.copy().multiply(0.2).addJ(0.1);

    updateDataCenter();
    this.personalBestPosition = this.position;
    this.personalBestObjectiveValue = this.dataCenter.computeObjective();
    // TODO remove
    this.personalBestMakespan = this.dataCenter.computeMakespan();
    this.personalBestThroughput = this.dataCenter.computeThroughput();
  }

  private void updateDataCenter() {
    this.dataCenter.resetVirtualMachineReadyTimes();

    for (int i = 0; i < this.position.getRowsCount(); i++) {
      Task task = workload.getTaskById(i);
      VirtualMachine virtualMachine = dataCenter.getVirtualMachineById((int) this.position.getIndexOfFirstNonZeroColumnForRow(i));
      this.dataCenter.addExecutionTimeToVirtualMachine(task, virtualMachine);
    }
  }

  private void updateVelocity(double w) {
    // Randomize constants r1 and r2
    double r1 = Utilities.getRandomDouble(0, 1);
    double r2 = Utilities.getRandomDouble(0, 1);

    // System.out.println("\nParticle " + this.id + " Velocity (before): " + this.velocity);
    Matrix previousVelocityFactor = this.velocity.copy().multiply(w);
    Matrix localExploration = this.personalBestPosition.copy().subtract(this.position).multiply(Particle.c1).multiply(r1);
    Matrix globalExploration = this.globalBestPosition.copy().subtract(this.position).multiply(Particle.c2).multiply(r2);
    this.velocity = previousVelocityFactor.add(localExploration).add(globalExploration)
                                          .enforceElementwiseBound(Particle.maxAbsoluteVelocity);
    // System.out.println("Particle " + this.id + " Velocity (after): " + this.velocity);
  }

  private void updatePosition() {
    // Instead of the standard PSO update equation, use the one defined in https://www.sciencedirect.com/science/article/pii/S1319157820305279#e0045
    // Because this version of PSO is discrete in nature
    // System.out.println("\nParticle " + this.id + " Position (before): " + this.position);
    this.position.zeroOut();  // TODO make this O(# tasks) instead of O(# tasks * # vms). Would this actually make a noticable difference? Because velocity is  O(# tasks * # vms)

    for (int i = 0; i < this.velocity.getRowsCount(); i++) {
      int j = this.velocity.getIndexOfMaximumColumnForRow(i);
      this.position.setComponent(i, j, 1);
    }
    // System.out.println("Particle " + this.id + " Position (after): " + this.position);
  }

  public int getId() {
    return this.id;
  }

  public Matrix getPosition() {
    return this.position;
  }

  public Matrix getVelocityVector() {
    return this.velocity;
  }

  private void updateTaskVmMapping() {
    this.taskVmMapping.clear();

    for (int i = 0; i < this.position.getRowsCount(); i++) {
      Task task = workload.getTaskById(i);
      VirtualMachine virtualMachine = dataCenter.getVirtualMachineById((int) this.position.getIndexOfFirstNonZeroColumnForRow(i));
      this.taskVmMapping.put(task, virtualMachine);
    }
  }

  public HashMap<Task,VirtualMachine> getTaskVmMapping() {
    this.updateTaskVmMapping();
    return this.taskVmMapping;
  }

  public void setGlobalBestPosition(Matrix globalBestPosition) {
    this.globalBestPosition = globalBestPosition;
  }

  public Matrix getPersonalBestPosition() {
    return this.personalBestPosition;
  }

  public double getPersonalBestObjectiveValue() {
    return this.personalBestObjectiveValue;
  }

  public double getPersonalBestMakespan() {
    return this.personalBestMakespan;
  }

  public double getPersonalBestThroughput() {
    return this.personalBestThroughput;
  }

  public ArrayList<Double> getObjectiveHistory() {
    return this.objectiveHistory;
  }

  @Override
  public String toString() {
    return "(Particle" + this.id + ")";
  }
}

File: psoscheduler-rust/src/pso/particle.rs
use ndarray::Array;
use ndarray::Array2;
use ndarray_rand::RandomExt;
use ndarray_rand::rand_distr::Uniform;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::collections::HashMap;

use crate::simulation::task::Task;
use crate::utils::utilities;
use crate::simulation::data_center::DataCenter;
use crate::simulation::workload::Workload;

static PARTICLE_ID_COUNTER: AtomicUsize = AtomicUsize::new(0);

pub struct Particle {
  pub id: usize,
  pub position: Array2<f32>,
  pub velocity: Array2<f32>,
  pub task_vm_mapping: HashMap<usize, usize>,
  pub personal_best_position: Array2<f32>,
  pub personal_best_objective: f32,
  pub personal_best_throughput: f32,
  pub objective_history: Vec<f32>,
}

impl Particle {
  const C1: f32 = 2.;
  const C2: f32 = 1.49455;

  const MAX_ABS_VELOCITY: f32 = 10.;

  pub fn new(workload: &mut Workload, data_center: &mut DataCenter) -> Particle {
    let mut position: Array2<f32> = Array::zeros((workload.tasks.len(), data_center.virtual_machines.len()));
    let task_vm_mapping: HashMap<usize, usize> = HashMap::new();

    data_center.reset_virtual_machine_ready_time();

    for task in workload.get_sorted_tasks().iter() {
      let vm_id: usize;
      if utilities::get_random_float(0., 1.) < 0.25 {
        vm_id = utilities::get_random_integer(0, data_center.virtual_machines.len());
      } else {
        vm_id = data_center.get_min_eet_virtual_machine(task);
      }
      position[[task.id, vm_id]] = 1.;
      
      // task_vm_mapping.insert(task.id, vm_id);
      data_center.add_execution_time_to_virtual_machine(task, vm_id);
    }

    let velocity:Array2<f32>  = Array::random((workload.tasks.len(), data_center.virtual_machines.len()), Uniform::new(0., 1.));
    let personal_best_position: Array2<f32> = position.clone();
    let personal_best_objective: f32 = data_center.compute_objective();
    let personal_best_throughput: f32 = data_center.compute_throughput();
    let objective_history: Vec<f32> = Vec::new();

    Particle {
      id: PARTICLE_ID_COUNTER.fetch_add(1, Ordering::Relaxed),
      position: position,
      velocity: velocity,
      task_vm_mapping: task_vm_mapping,
      personal_best_position: personal_best_position,
      personal_best_objective: personal_best_objective,
      personal_best_throughput: personal_best_throughput,
      objective_history: objective_history,
    }
  }

  pub fn run_iteration(&mut self, workload: &Workload, data_center: &mut DataCenter, w: f32, global_best_position: &Array2<f32>) -> u8 {
    self.update_velocity(w, global_best_position);
    self.update_position_and_data_center(workload, data_center);

    // self.update_data_center(workload, data_center);
    let objective: f32 = data_center.compute_objective();

    if objective > self.personal_best_objective {
      self.personal_best_objective = objective;
      self.personal_best_throughput = data_center.compute_throughput();
      self.personal_best_position = self.position.clone();
      self.objective_history.push(objective);
      return 1;
    }

    self.objective_history.push(self.personal_best_objective);

    return 0;
  }
  
  fn update_data_center(&self, workload: &Workload, data_center: &mut DataCenter) {
    data_center.reset_virtual_machine_ready_time();

    let mut i: usize = 0;
    for row in self.position.rows() {
      let task: &Task = &workload.tasks[i];
      let mut vm_id: usize = 0;
      for (j, e) in row.indexed_iter() {
        if *e > 0.9 {
          vm_id = j;
          break;
        }
      }
      data_center.add_execution_time_to_virtual_machine(task, vm_id);
      i += 1;
    }
  }

  fn update_velocity(&mut self, w: f32, global_best_position: &Array2<f32>) {
    let r1: f32 = utilities::get_random_float(0., 1.);
    let r2: f32 = utilities::get_random_float(0., 1.);

    self.velocity.mapv_inplace(|a| a * w);
    let mut local_exploration: Array2<f32> = &self.personal_best_position - &self.position;
    local_exploration.mapv_inplace(|a| a * Particle::C1 * r1);
    let mut global_exploration: Array2<f32> = global_best_position - &self.position;
    global_exploration.mapv_inplace(|a| a * Particle::C2 * r2);
    self.velocity += &(local_exploration + &global_exploration);
    self.velocity.mapv_inplace(|a| -> f32 {
      if a.abs() > Particle::MAX_ABS_VELOCITY || a < 0. {
        return utilities::get_random_float(0., Particle::MAX_ABS_VELOCITY);
      }
      return a;
    });
  }

  fn update_position_and_data_center(&mut self, workload: &Workload, data_center: &mut DataCenter) {
    data_center.reset_virtual_machine_ready_time();

    let mut i: usize = 0;
    for row in self.velocity.rows() {
      let mut max_col: usize = 0;
      let mut max_val: f32 = -Particle::MAX_ABS_VELOCITY - 1.;
      for (j, e) in row.indexed_iter() {
        self.position[[i, j]] = 0.;
        if *e > max_val {
          max_col = j;
          max_val = *e;
        }
      }
      self.position[[i, max_col]] = 1.;
      data_center.add_execution_time_to_virtual_machine(&workload.tasks.get(i).unwrap(), max_col);
      i += 1
    }
  }

  pub fn update_task_vm_mapping(&mut self) {
    self.task_vm_mapping.clear();
    let mut i: usize = 0;
    for row in self.position.rows() {
      for (j, e) in row.indexed_iter() {
        if *e > 0.9 {
          self.task_vm_mapping.insert(i, j);
        }
      }
      i += 1;
    }
  }
}
